### DISCUSSION TITLE
Suppress Object Level Business rules (not associated with properties)

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3671] [Questions] [answered]
We have three methods of suppressing property level rules (as discussed on page 92 of Using Csla objects book).
 From the UI, supressing rules only works after object has already been created such as
```cs
 ((Csla.Core.ICheckRules)_customer).SuppressRuleChecking();
 try
 {
 // set properties here without executing
 // any rules
 }
 finally
 {
 ((Csla.Core.ICheckRules)_customer).ResumeRuleChecking();
 ((Csla.Core.ICheckRules)_customer).CheckRules();
 }

```

We found out that when we create the object by calling DataPortal.Create, object levels rules are executed while property rules are added but not 'executed'.
Is there a way we could suppress object-level business rules?

Kind Regards

**Comment by Chicagoan2016** [comment]
That was easy : )
'StopIfIsNew' in RuleTutorial sample did the trick.

**Reply by Chicagoan2016** [reply]
I take it back lol.
I guess we still need something to suppress object rules during 'Create' operation.

**Reply by Chicagoan2016** [reply]
Or if we could 'execute'  property rules during 'Create'?

**Reply by ossendorf-at-hoelscher** [reply]
At which state are you in CSLA? In your `[Create]` method or where?

**Reply by Chicagoan2016** [reply]
As soon as I call Create, object rules are 'executed' but propertyrules aren't. Property rules are added though (in AddBusinessRules())

**Comment by rockfordlhotka** [comment, marked as answer]
Here's a basic test on which we can build a discussion. I'm not sure where you are suppressing rules?

If the `CheckRules` method is not calling in the create operation, no rules run. if it is called in the create operation then the type and property rules run.


```c#
using Csla;
using Csla.Configuration;
using Csla.Rules;
using Microsoft.Extensions.DependencyInjection;

var services = new ServiceCollection();
services.AddCsla();
var provider = services.BuildServiceProvider();
var portal = provider.GetRequiredService<IDataPortal<PersonEdit>>();
var obj = portal.Create();


[Serializable]
public class PersonEdit : BusinessBase<PersonEdit>
{
    public static readonly PropertyInfo<string> NameProperty = RegisterProperty<string>(nameof(Name));
    public string Name
    {
        get => GetProperty(NameProperty);
        set => SetProperty(NameProperty, value);
    }

    protected override void AddBusinessRules()
    {
        base.AddBusinessRules();
        BusinessRules.AddRule(new TypeRule());
        BusinessRules.AddRule(new PropertyRule { PrimaryProperty = NameProperty });
    }

    [Create]
    private void Create()
    {
        Console.WriteLine(nameof(Create));
        //BusinessRules.CheckRules();
    }
}

public class TypeRule : BusinessRule
{
    protected override void Execute(IRuleContext context)
    {
        Console.WriteLine(nameof(TypeRule));
    }
}

public class PropertyRule : BusinessRule
{
    protected override void Execute(IRuleContext context)
    {
        Console.WriteLine(nameof(PropertyRule));
    }
}
```

**Comment by Chicagoan2016** [comment]
@rockfordlhotka Thank you Sir. 
Commenting out the line CheckRules in Create did the trick.
We have some old applications in Csla 4.xx, do we have to comment out the line 
base.DataPortal_Create();
 
to get the same behavior as above?

Kind Regards

**Reply by rockfordlhotka** [reply]
I would think so.

**Comment by Chicagoan2016** [comment]
@rockfordlhotka  I noticed that if we assign a priority of -10 to PropertyRule  and a priority of 10 to TypeRule, TypeRule is still executed first (if we don't comment out the line BusinessRules.CheckRules(); in Create).

Kind Regards

**Reply by rockfordlhotka** [reply]
iirc, type rules and property rules are run entirely separately - not mixed together by priority.

In fact, I'm not entirely sure type rules use priority at all, but I'm not sure.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka  Thank you Sir.
What would be a good strategy to follow if we need to make sure certain rules are executed first? In some objects, we have both property and type rules.
Regards

**Reply by rockfordlhotka** [reply]
I'd have to understand the specific goals and requirements to provide a meaningful answer.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka When I assign primary property to Typerule, the 'Priority' works as expected : )

**Reply by Chicagoan2016** [reply]
@rockfordlhotka  As an 80/'90s kid, this reminds me of  Perfect Strangers' Balki Bartokomous. He would say 'We are so happy, we do the dance of joy' 😂
